pkg/client/client.go

```diff
@@ -205,21 +208,104 @@ func NewClientFromKeys(accessKey, secretAccessKey, region string) (Client, error
 	}, nil
 }
 
+// DescribeRegionsData holds output of DescribeRegions API call and the time when it was last updated.
+type DescribeRegionsData struct {
+	err                   error
+	describeRegionsOutput *ec2.DescribeRegionsOutput
+	lastUpdated           time.Time
+}
+
+// cachedDescribeRegions returns full list of regions from AWS.
+// DescribeRegionsOutput is cached to avoid AWS API calls on each reconcile loop.
+func cachedAWSDescribeRegions(awsSession *session.Session, regionsCacheMap map[string]DescribeRegionsData) (*ec2.DescribeRegionsOutput, error) {
+	creds, err := awsSession.Config.Credentials.Get()
+	if err != nil {
+		return nil, err
+	}
+	accessKeyID := creds.AccessKeyID
+	regionsCache, found := regionsCacheMap[accessKeyID]
+	if !found {
+		regionsCache = DescribeRegionsData{}
+	}
+
+	if regionsCache.describeRegionsOutput != nil && regionsCache.err == nil &&
+		time.Since(regionsCache.lastUpdated) < awsRegionsCacheExpirationDuration {
+		klog.Info("Using cached AWS region data")
+		return regionsCache.describeRegionsOutput, regionsCache.err
+	}
+
+	currentRegion := awsSession.Config.Region
+	// Use known endpoint to send our request
+	awsSession.Config.Region = aws.String("us-east-1")
```

Comment:
Is it possible that a user may not be allowed to make this describe regions request for a particular region? How does the AWS IAM work here? 🤔

Reply:
The permission is global. You can use any region for this API. us-east-1 is the default region that is used when you don't specify region. Changed this to override endpoint instead of region. I am now using global "https://ec2.amazonaws.com" endpoint, so that it can go  anywhere AWS wants. (Currently it queries us-east-1)

Reply:
Okay, this does not work with the global endpoint. Probably forgot to hit save/rebuild when testing this. Rolled back to changing the region to us-east-1

Reply:
Ok let's see how this goes